### PR TITLE
ByteProfile: an auto profiling tool for BytePS

### PR DESCRIPTION
**Pull request by joapolarbear**
An auto profiling tool integrated into byteps. Currently, it only supports mxnet. Auto profiling tools for  Tensorflow and Pytorch is in developing.

## Features
* Support both Module API and Gluon API
* Output traces in the standard Chrome Trace format, which can be further visualized as a timeline through [chrome](chrome://tracing). An example of the timeline is shown below.
* We focus on each operation (including forward OPs, backward OPs, and communication OPs) of MXNet, instead of kernels, because the traces of operations is more useful for us to analyze the bottleneck of a distributed ML job.
* For the analysis of traces, we also profile the dependencies between these operations.
* In byteps, when synchronizing a gradient with the PS, one gradient is divided into multiple partitions and each partition needs to go through multiple types of operations, e.g., `REDUCE`, `PUSH`, `PULL`, `BROADCAST`. Byteprofile allows us to see the traces for each of these operations.

byteps_timeline_mnist_2WK_2V100GPU

## Detailed
To learn how to use this profiling tool, please refer to [README.md](https://github.com/joapolarbear/byteps/blob/byteprofile/example/mxnet/README.md).

## TODO
* Auto profiling tools for  Tensorflow and Pytorch is in developing.
* Automatically analyze the traces and find the bottleneck.

### CODE REVIEW

**Comment by ymjiang** [comment]
Thank you for the contribution. I leave some preliminary reviews on the code format. And we definitely need more time to look into the code logic. So far I have two questions:
1. If we disable the profiling, is the training performance as normal as before?
2. If we enable the profiling, how much degree will the performance drop?

**Comment by bobzhuyb** [comment]
Hi @joapolarbear , I understand that this timeline implementation can work with MXNet profiler. However, BytePS itself is cross-framework and not tied to just MXNet. I am wondering whether we can get a clean timeline implementation just focused on BytePS itself? The MXNet-specific part should target towards MXNet upstream.

**Comment by bobzhuyb** [comment]
Hi @joapolarbear , should we close this PR for now. Once you are ready, we can reopen this. Otherwise, github is sending too many notifications about every commit you make..

**Comment by joapolarbear** [comment]
Sure.

Yibo Zhu <notifications@github.com> 于2019年11月18日周一 上午10:58写道：
